File: samples/java/app/src/test/java/org/remdev/wetweexapp/TestUtils.java
package org.remdev.wetweexapp;

import androidx.core.util.Consumer;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import org.mockito.Mockito;
import org.remdev.wetweex.component.Coordinator;
import org.remdev.wetweex.live.TweexMutableLiveData;
import org.remdev.wetweex.model.LoadProgressInfo;
import org.remdev.wetweex.viewmodel.TweexProgressViewModel;
import org.remdev.wetweex.viewmodel.TweexViewModel;

import java.util.concurrent.CountDownLatch;

public class TestUtils {

    public static <T extends TweexViewModel> CountDownLatch waitCoordinatorLatch(Coordinator<T> coordinator) {
        final CountDownLatch latch = new CountDownLatch(1);
        coordinator.addTasksCompletedListener(latch::countDown);
        return latch;
    }

    public static <T extends TweexViewModel, C extends Coordinator<T>> void runCoordinator(C coordinator,
                                                                                           Consumer<C> operation) {
        final CountDownLatch latch = waitCoordinatorLatch(coordinator);
        operation.accept(coordinator);
        try {
            latch.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public static <T> T mockViewModel(Class<T> cls) {
        T viewModel = Mockito.mock(cls);
        if (viewModel instanceof TweexProgressViewModel) {
            appendProgressInfoMock((TweexProgressViewModel) viewModel);
        }
        return viewModel;
    }

    private static void appendProgressInfoMock(TweexProgressViewModel viewModel) {
        final TweexMutableLiveData mockLoad = new TweexMutableLiveData<LoadProgressInfo>() {

            private LoadProgressInfo data = new LoadProgressInfo();

            @Override
            public void setOrPost(LoadProgressInfo loadProgressInfo) {
                this.data = data;
            }

            @Override
            public void observe(LifecycleOwner lifecycleOwner, Observer<? super LoadProgressInfo> observer) {

            }

            @Override
            public LoadProgressInfo getValue() {
                return data;
            }
        };
        Mockito.when(viewModel.getLoadingInfo()).thenReturn(mockLoad);
    }
}

File: samples/kotlin/app/src/test/java/org/remdev/wetweexapp/TestUtils.kt
package org.remdev.wetweexapp

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import org.remdev.wetweex.component.Coordinator
import org.remdev.wetweex.live.TweexMutableLiveData
import org.remdev.wetweex.model.LoadProgressInfo
import org.remdev.wetweex.viewmodel.TweexProgressViewModel
import org.remdev.wetweex.viewmodel.TweexViewModel
import org.mockito.Mockito
import java.util.concurrent.CountDownLatch


fun <T : TweexViewModel> waitCoordinatorLatch(coordinator: Coordinator<T>): CountDownLatch {
    val latch = CountDownLatch(1)
    coordinator.addTasksCompletedListener {
        latch.countDown()
    }
    return latch
}

fun <T : TweexViewModel> runCoordinator(coordinator: Coordinator<T>, operation: (Coordinator<T>) -> Unit) {
    val latch = waitCoordinatorLatch(coordinator)
    operation.invoke(coordinator)
    return latch.await()
}

inline fun <reified T> mockViewModel(): T {
    val viewModel = Mockito.mock(T::class.java)
    if (viewModel is TweexProgressViewModel) {
        appendProgressInfoMock(viewModel)
    }
    return viewModel
}

fun appendProgressInfoMock(viewModel: TweexProgressViewModel) {
    val mockLoad = object : TweexMutableLiveData<LoadProgressInfo> {
        private var data: LoadProgressInfo = LoadProgressInfo()

        override fun observe(owner: LifecycleOwner?, data: Observer<in LoadProgressInfo>?) {}

        override fun getValue(): LoadProgressInfo = data

        override fun setOrPost(data: LoadProgressInfo) {
            this.data = data
        }
    }
    Mockito.`when`(viewModel.loadingInfo).thenReturn(mockLoad)
}